projects/graph/graph.py

```diff
@@ -16,42 +15,66 @@ def add_vertex(self, vertex):
         """
         Add a vertex to the graph.
         """
-        pass  # TODO
+        self.vertices[vertex] = set()
     def add_edge(self, v1, v2):
         """
         Add a directed edge to the graph.
         """
-        if v1 not self.vertices:
+        if v1 not in self.vertices:
             self.add_vertex(v1);
+
         if v2 not in self.vertices:
             self.add_vertex(v2);
-        self.vertices[v1].add(v2);
+
+        self.vertices[v1].add(v2)
         
         
     def bft(self, starting_vertex):
         """
         Print each vertex in breadth-first order
         beginning from starting_vertex.
         """
-       #make a queue
-       queue = Queue()
-       #make a visited set
-       visited = set()
-       # put starting vertex in he queue
-       queue.enqueue(starting_vertex)
-       #while q isn\t empty
-       while queue.size():
-       # dequeue the item, it is our current item
+        #make a queue
+        queue = Queue()
+        #make a visited set
+        visited = set()
+        # put starting vertex in he queue
+        queue.enqueue(starting_vertex)
+        #while q isn\t empty
+        while queue.size():
+        # dequeue the item, it is our current item
             node = queue.dequeue()
             print(node)
-       # mark current as visited
-            visited.add
+        # mark current as visited
+            visited.add(node)
+        # for each of the dequeued item's edge
+            for edge in self.vertices[node]:
+        # put them in the queue
+                queue.enqueue(edge)
     def dft(self, starting_vertex):
         """
         Print each vertex in depth-first order
         beginning from starting_vertex.
         """
-        pass  # TODO
+        # make a stack
+        stack = Stack()
+        # make a visited set
+        visited = set()
+        # put starting vertex in our stack
+        stack.push(starting_vertex)
+        #while the stack isn't empty
+        while stack.size():
+            ## pop off the top of the stack, it is our current item
```

Comment:
Popping an item from a stack will result in the last item from being removed from the stack, instead of the top or first item.